### DISCUSSION TITLE
Rewrite of HTTPS Requests

### DISCUSSION THREAD

**Discussion by ChrisJBurns** [coredns/coredns#6215] [Q&A]
I currently have the following ConfigMap for CoreDNS in use on EKS.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    eks.amazonaws.com/component: coredns
    k8s-app: kube-dns
    managed-by: flux-infra
  name: coredns
  namespace: kube-system
data:
  Corefile: |-
    .:53 {
        errors
        health
        rewrite name auth.test.our.domain auth-idp.test.svc.cluster.local
        rewrite name auth.test.our.domain-dev.uk auth-idp.test.svc.cluster.local
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

However when I go into any of the pods and make the following request:
`curl -I https://auth.test.our.domain -k` it hangs and eventually fails. `curl -I http://auth.test.our.domain` works completely fine. And when I remove the 2 rewrite rules above, the https curl request also then works. So was wondering if we needed to add something extra to allow for the HTTPS request to work?

**Comment by chrisohaver** [comment]
It could be query name mismatch and old versions of coredns do not automatically do answer rewrites for name rewrites.  I don't recall what version it was added.  Check the section on answer rewrites in the rewrite plugin readme.

Also you shouldn't need specify the same rule twice. The copy just gets ignored.

**Reply by ChrisJBurns** [reply]
I'll check the rewrite plugin README. I've amended the code sample to factor in the difference in the second URL as there is a slight difference.

**Reply by chrisohaver** [reply]
Oh - nevermind... http vs https ...  I'm guessing this is a cert signing issue, the cert is probably signed to auth-idp.test.svc.cluster.local, resulting in cert name mismatch.

**Reply by ChrisJBurns** [reply]
Am running `coredns:v1.7.0-eksbuild.1` for reference, which I believe is one of the latest (if not the latest).

Cert issue does sound plausible given the endpoint is signed with a self-signed cert. I'm imagining I just need to insert the cert into the CoreDNS `/etc/ssl/certs` directory?

**Reply by chrisohaver** [reply]
I mean the cert is issued for `auth-idp.test.svc.cluster.local`, and curl is going to a url named `auth.test.our.domain`, the cert that curl receives is for `auth-idp.test.svc.cluster.local`, not the url it is trying to reach `auth.test.our.domain`.

curl: "That's somebody else's cert"

**Reply by ChrisJBurns** [reply]
Would CoreDNS be issuing the cert for `auth-idp.test.svc.cluster.local`?

**Reply by chrisohaver** [reply]
No, the cert exchange occurs after the DNS part of the connection. The cert comes from auth-idp.test.svc.cluster.local.

**Reply by chrisohaver** [reply]
All DNS is doing is providing an IP for the hostname. Once dns provides this, its job is done.

1. curl wants to connect to "auth.test.our.domain", so it asks dns for an IP address for auth.test.our.domain.
2. dns returns an IP address. behind the scenes, coredns is actually retrieving the ip for auth-idp.test.svc.cluster.local, but curl has no knowledge of this. curl thinks the ip received is for auth.test.our.domain.
3. curl then connects to that IP and gets a ssl cert.  curl, verifies the server's identity in the cert (it also verifies that the cert is signed by a trusted signer). The identity check would fail because the cert identifies the server as "auth-idp.test.svc.cluster.local", and curl is expecting "auth.test.our.domain".

**Reply by ChrisJBurns** [reply]
Thanks for the extra info!

I guess I'm trying to deduce what's issuing the cert for `auth-idp.test.svc.cluster.local`. Even when I pass `-k` to curl which should bypass cert validation, it gives connection reset by peer errors. 

When running `openssl s_client -connect auth-idp.test:443 -showcerts` I just get
```
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 316 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
```
